pkg/controller/direct/compute/forwardingrule_controller.go

```diff
@@ -0,0 +1,482 @@
+	globalForwardingRulesClient, err := gcp.NewGlobalForwardingRulesRESTClient(ctx, opts...)
+	if err != nil {
+		return nil, fmt.Errorf("building global ComputeForwardingRule client: %w", err)
+	}
+	return globalForwardingRulesClient, err
+}
+
+func (m *forwardingRuleModel) forwardingRulesClient(ctx context.Context) (*gcp.ForwardingRulesClient, error) {
+	var opts []option.ClientOption
+	if m.config.UserAgent != "" {
+		opts = append(opts, option.WithUserAgent(m.config.UserAgent))
+	}
+	if m.config.HTTPClient != nil {
+		opts = append(opts, option.WithHTTPClient(m.config.HTTPClient))
+	}
+	if m.config.UserProjectOverride && m.config.BillingProject != "" {
+		opts = append(opts, option.WithQuotaProject(m.config.BillingProject))
+	}
+
+	forwardingRulesClient, err := gcp.NewForwardingRulesRESTClient(ctx, opts...)
+	if err != nil {
+		return nil, fmt.Errorf("building ComputeForwardingRule client: %w", err)
+	}
+	return forwardingRulesClient, err
+}
+
+func (m *forwardingRuleModel) AdapterForObject(ctx context.Context, reader client.Reader, u *unstructured.Unstructured) (directbase.Adapter, error) {
+	obj := &krm.ComputeForwardingRule{}
+	if err := runtime.DefaultUnstructuredConverter.FromUnstructured(u.Object, &obj); err != nil {
+		return nil, fmt.Errorf("error converting to %T: %w", obj, err)
+	}
+
+	// Get ResourceID
+	resourceID, err := refs.GetResourceID(u)
+	if err != nil {
+		return nil, err
+	}
+
+	// Get projectID
+	projectID, err := refs.ResolveProjectID(ctx, reader, u)
+	if err != nil {
+		return nil, err
+	}
+
+	// Get network
+	if obj.Spec.NetworkRef != nil {
+		networkRef, err := refs.ResolveComputeNetwork(ctx, reader, obj, obj.Spec.NetworkRef)
+		if err != nil {
+			return nil, err
+
+		}
+		obj.Spec.NetworkRef.External = networkRef.ID()
+	}
+
+	// Get compute address
+	if obj.Spec.IpAddress.AddressRef != nil {
+		computeAddressRef, err := refs.ResolveComputeAddress(ctx, reader, obj, obj.Spec.IpAddress.AddressRef)
+		if err != nil {
+			return nil, err
+
+		}
+		obj.Spec.IpAddress.AddressRef.External = computeAddressRef.GetAddress()
+	}
+
+	// Get target targetHTTPProxy
+	if obj.Spec.Target.TargetHTTPProxyRef != nil {
+		targetHTTPProxyRef, err := refs.ResolveTargetHTTPProxy(ctx, reader, obj, obj.Spec.Target.TargetHTTPProxyRef)
+		if err != nil {
+			return nil, err
+
+		}
+		obj.Spec.Target.TargetHTTPProxyRef.External = targetHTTPProxyRef.URL()
+	}
+
+	// Get target TargetVPNGateway
+	if obj.Spec.Target.TargetVPNGatewayRef != nil {
+		targetVPNGatewayRef, err := refs.ResolveComputeTargetVPNGateway(ctx, reader, obj, obj.Spec.Target.TargetVPNGatewayRef)
+		if err != nil {
+			return nil, err
+
+		}
+		obj.Spec.Target.TargetVPNGatewayRef.External = targetVPNGatewayRef.URL()
+	}
+
+	// Get location
+	location := obj.Spec.Location
+
+	forwardingRuleAdapter := &forwardingRuleAdapter{
+		resourceID: resourceID,
+		projectID:  projectID,
+		Location:   location,
+		desired:    obj,
+	}
+
+	// Get compute GCP client
+	if location == "global" {
```

Comment:
This works for me! I think your approach is good enough for now and easy to maintain (which is really important).